pmlc/dialect/tile/transforms/constant_types.cc

```diff
@@ -0,0 +1,121 @@
+using mlir::OpRewritePattern;
+using mlir::OwningRewritePatternList;
+using mlir::PatternMatchResult;
+using mlir::PatternRewriter;
+
+using pmlc::dialect::eltwise::ScalarConstantOp;
+
+static llvm::cl::OptionCategory constant_types_options("tile-constant-types options");
+
+static llvm::cl::opt<std::string> constant_types_option_floatx("tile-constant-types-floatx", llvm::cl::init("float32"),
+                                                               llvm::cl::desc("set floating-point constant precision"),
+                                                               llvm::cl::cat(constant_types_options));
+
+static llvm::cl::opt<std::string> constant_types_option_intx("tile-constant-types-intx", llvm::cl::init("int32"),
+                                                             llvm::cl::desc("set integer constant precision"),
+                                                             llvm::cl::cat(constant_types_options));
+
+struct ConstantTypesRewriter : public OpRewritePattern<ScalarConstantOp> {
+  ConstantTypesRewriter(mlir::MLIRContext* context, DataType floatx, DataType intx)
+      : OpRewritePattern<ScalarConstantOp>(context), floatx_(floatx), intx_(intx) {}
+
+  DataType floatx_;
+  DataType intx_;
+
+  PatternMatchResult matchAndRewrite(ScalarConstantOp constOp, PatternRewriter& rewriter) const override;
+};
+
+PatternMatchResult ConstantTypesRewriter::matchAndRewrite(ScalarConstantOp constOp, PatternRewriter& rewriter) const {
+  IVLOG(3, "ConstantTypesPass::matchAndRewrite");
+
+  auto type = constOp.getType();
+  auto shape = eltwise::getRankedTensorType(type).getShape();
+
+  auto float_attr = constOp.getFloatAttr();
+  auto int_attr = constOp.getIntAttr();
+
+  if (float_attr && floatx_ != DataType::invalid) {
+    double value = float_attr.getValueAsDouble();
+    auto elementType = ScalarType::get(type.getContext(), floatx_);
+    auto new_type = RankedTensorType::get(shape, elementType);
+
+    if (new_type == type) {
+      return matchFailure();
+    }
+    rewriter.replaceOpWithNewOp<ScalarConstantOp>(constOp, elementType, value);
+    return matchSuccess();
+
+  } else if (int_attr && intx_ != DataType::invalid) {
+    int64_t value = int_attr.getInt();
+
+    if (pmlc::util::isUnsigned(intx_) && (value < 0)) {
+      std::stringstream ss;
+      ss << "Invalid datatype for negative constant";
+      throw std::runtime_error(ss.str());
+    }
+    auto elementType = ScalarType::get(type.getContext(), intx_);
+    auto new_type = RankedTensorType::get(shape, elementType);
+
+    if (new_type == type) {
+      return matchFailure();
+    }
+    rewriter.replaceOpWithNewOp<ScalarConstantOp>(constOp, elementType, value);
+    return matchSuccess();
+  } else {
+    IVLOG(1, "Warning: non-float, non-int type");
+  }
+  return matchFailure();
+}
+
+struct ConstantTypesPass : public OperationPass<ConstantTypesPass> {
+  ConstantTypesPass(const std::string& floatx = constant_types_option_floatx,
```

Comment:
Use `llvm::StringRef` instead of `const std::string&`.

Reply:
Actually, could we make it so that the `ConstantTypesPass` always takes in a `DataType`, and the conversion between string to int happens as part of the command line parsing code.